test/dynamo/test_activation_checkpointing.py

```diff
@@ -324,13 +391,292 @@ def fn(x, y, z):
+                _get_default_policy(no_recompute_list=no_recompute_list), storage
+            )
+            return caching_mode, cached_mode
+
+        def gn(x, y):
+            return torch.sigmoid(torch.selu(torch.matmul(torch.matmul(x, y), y)))
+
+        def fn(x, y):
+            return torch.utils.checkpoint.checkpoint(
+                gn,
+                torch.sin(x),
+                y,
+                use_reentrant=False,
+                context_fn=selective_checkpointing_context,
+            )
+
+        x = torch.randn(4, 4, device="cuda", requires_grad=True)
+        y = torch.randn(4, 4, device="cuda", requires_grad=True)
+
+        fw_compiler = functools.partial(
+            count_ops_list,
+            freqs=[2, 1],
+            ops=[torch.ops.aten.mm.default, torch.ops.aten.sigmoid.default],
+        )
+        bw_compiler = functools.partial(
+            count_ops_list,
+            freqs=[4, 0],
+            ops=[torch.ops.aten.mm.default, torch.ops.aten.sigmoid.default],
+        )
+        backend = aot_autograd(fw_compiler=fw_compiler, bw_compiler=bw_compiler)
+        self._validate(fn, backend, x, y)
+
+
+    @requires_cuda()
+    def test_tags_selective_checkpoint_inplace_op(self):
+        def selective_checkpointing_context():
+            no_recompute_list = [
+                torch.ops.aten.mm.default,
+                torch.ops.aten.sigmoid.default,
+            ]
+            storage = []
+            caching_mode = CachingTorchDispatchMode(
+                _get_default_policy(no_recompute_list=no_recompute_list), storage
+            )
+            cached_mode = CachedTorchDispatchMode(
+                _get_default_policy(no_recompute_list=no_recompute_list), storage
+            )
+            return caching_mode, cached_mode
+
+        def gn(x, y):
+            return torch.sigmoid(torch.selu_(torch.matmul(torch.matmul(x, y), y)))
+
+        def fn(x, y):
+            return torch.utils.checkpoint.checkpoint(
+                gn,
+                torch.sin(x),
+                y,
+                use_reentrant=False,
+                context_fn=selective_checkpointing_context,
+            )
+
+        x = torch.randn(4, 4, device="cuda", requires_grad=True)
+        y = torch.randn(4, 4, device="cuda", requires_grad=True)
+
+        fw_compiler = functools.partial(
+            count_ops_list,
+            freqs=[2, 1],
+            ops=[torch.ops.aten.mm.default, torch.ops.aten.sigmoid.default],
+        )
+        bw_compiler = functools.partial(
+            count_ops_list,
+            freqs=[4, 0],
+            ops=[torch.ops.aten.mm.default, torch.ops.aten.sigmoid.default],
+        )
+        backend = aot_autograd(fw_compiler=fw_compiler, bw_compiler=bw_compiler)
+        with self.assertRaisesRegex(
+            AssertionError,
+            "In-place ops are not supported in selective checkpointing region under torch.compile"
+        ):
+            self._validate(fn, backend, x, y)
```

Comment:
@Chillee I compared this test case with `test_tags_selective_checkpoint_outplace_op` above (`torch.selu_` vs. `torch.selu`), and it seems that as long as we have in-place op applied to inputs or activations in the graph, the gradient comparison check would fail.

Reply:
Hmm... I need to think a bit more about this. I would ... expect that functionalization would apply underneath selective checkpointing, and so it would still work correctly.

Reply:
Hmm. This seems pretty similar to the "what should run first, torch_dispatch vs. functionalization" ordering issue I'm running into with subclass support. I don't think this will work today, since the "default" thing that happens is that functionalization has a higher priority in the dispatcher than __toch_dispatch__.

I updated the design doc after running into some issues with DTensor - I think functionalization will need to run below torch_dispatch (updated [here](https://docs.google.com/document/d/1COQ291nOZvtFnzGTQMJqoYZ3sttEYFw_7HbfSyL8gcA/edit#bookmark=id.4tvs5sten7fb)). Hopefully that should fix this problem too - I can give a shout once I have a working PR to test with if that sounds good to you all.

Reply:
But (once the functionalization ordering issue is sorted out), agreed with Horace's comment - instead of manually checking for inplace ops, would it make sense to use the existing `speculate_graph` infra that Voz/Richard wrote, that just tries to ban subgraphs containing input mutations?

Reply:
yes - I can confirm that with Brian's mode reordering stack, I can use `func._schema.is_mutable` to check inplace ops.

Reply:
Hmm @yf225 do we want to ban **all** mutable ops for now? Or is that needed any time soon.

Functionalization should handle any mutable ops inside the checkpointed region. The main issue is just "what if a mutable op tries to mutate one of the **inputs** to the checkpointed region" - and in that case, higher order ops today are supposed to error in `speculate_subgraph` if they detect that there were any input mutations.

Reply:
Update: will wait for Brian's stack to land for enabling in-place op support